### PR TITLE
[new release] tls, tls-mirage, tls-lwt, tls-eio and tls-async (0.16.0)

### PR DESCRIPTION
**Pull request by hannesm**
Transport Layer Security purely in OCaml

- Project page: <a href="https://github.com/mirleft/ocaml-tls">https://github.com/mirleft/ocaml-tls</a>
- Documentation: <a href="https://mirleft.github.io/ocaml-tls/doc">https://mirleft.github.io/ocaml-tls/doc</a>

##### CHANGES:

* BREAKING: new opam package tls-lwt (formerly tls.lwt), in dune:
  (libraries tls.lwt) should now be libraries (tls-lwt)
  (mirleft/ocaml-tls#468 @hannesm, reported mirleft/ocaml-tls#449 by @mbacarella)
* tls: update to mirage-crypto 0.11 API (mirleft/ocaml-tls#468 @hannesm)
* tls: relax SignatureAlgorithms extension handling to allow OpenSSL
  interoperability tests with TLS 1.0 and TLS 1.1 (mirleft/ocaml-tls#469 @hannesm)
* tls: remove Utils.filter_map and and Utils.option, use Stdlib instead (mirleft/ocaml-tls#455
  @hannesm)
* tls: do not globally open Utils (mirleft/ocaml-tls#455 @hannesm)
* tls: export log source of Tracing module (mirleft/ocaml-tls#461 @bikallem)
* tls: remove unused ciphersuites to reduce binary size (mirleft/ocaml-tls#467 @hannesm)
* tls-lwt: do not catch out of memory exception (mirleft/ocaml-tls#469 @hannesm)
* tls-eio: add fuzz testing using crowbar (mirleft/ocaml-tls#456 mirleft/ocaml-tls#463 @talex5)
* tls-eio: update to eio 0.7 (mirleft/ocaml-tls#456 @talex5)
* tls-eio: fix test for develop with vendoring (mirleft/ocaml-tls#462 @bikallem)

### CODE REVIEW

**Comment by hannesm** [comment]
Since ocamlfind's `tls.lwt` is now `tls-lwt`, there's expectation of failures (in addition to the API changes in #23282). I'll observe the CI and add in separate PR(s) upper bound constraints (of course, anyone helping out is highly appreciated).

**Comment by hannesm** [comment]
I have a question for both the opam-repository maintainers and @talex5: the tls-async.opam has a "ocaml" {< "5.0.0"} bound, and I understand this was useful at some point for some CI system(s) -- but is this the right thing to have? I do not think that anything in tls-async is incompatible with OCaml 5.0. Any advice on this is highly appreciated (I test the relaxation at https://github.com/mirleft/ocaml-tls/pull/471).

**Comment by hannesm** [comment]
update: I removed the upper bound for ocaml

**Comment by talex5** [comment]
> I have a question for both the opam-repository maintainers and @talex5: the tls-async.opam has a "ocaml" {< "5.0.0"} bound, and I understand this was useful at some point for some CI system(s) -- but is this the right thing to have?

https://github.com/mirleft/ocaml-tls/commit/ea1631c19472084ba0cd5d3d6c143da63483ca6e says:
> * Temporary fix for tls-async in CI
> tls-async can't be tested on OCaml 5 yet, and this prevents the CI from
testing other packages there. This annotation can be removed once a
compatible version of `core` is released.

So it should be fine to remove it now (v0.15.1 seems to work with OCaml 5).

File: packages/tls-eio/tls-eio.0.16.0/opam [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+bug-reports:  "https://github.com/mirleft/ocaml-tls/issues"
+doc:          "https://mirleft.github.io/ocaml-tls/doc"
+authors:      ["Thomas Leonard"]
+maintainer:   ["Hannes Mehnert <hannes@mehnert.org>" "David Kaloper <david@numm.org>"]
+license:      "BSD-2-Clause"
+
+build: [
+  ["dune" "subst"] {dev}
+  ["dune" "build" "-p" name "-j" jobs]
+  ["dune" "runtest" "-p" name "-j" jobs] {with-test}
```
Comment (kit-ty-kate) [inline packages/tls-eio/tls-eio.0.16.0/opam:13]:
```suggestion
  ["dune" "runtest" "-p" name "-j" jobs] {with-test & os != "macos"}
```
```

#=== ERROR while compiling tls-eio.0.16.0 =====================================#
# context              2.1.4 | macos/x86_64 | ocaml-base-compiler.5.0.0 | pinned(https://github.com/mirleft/ocaml-tls/releases/download/v0.16.0/tls-0.16.0.tbz)
# path                 ~/.opam/5.0.0/.opam-switch/build/tls-eio.0.16.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune runtest -p tls-eio -j 11
# exit-code            1
# env-file             ~/.opam/log/tls-eio-92405-ec74b4.env
# output-file          ~/.opam/log/tls-eio-92405-ec74b4.out
### output ###
# (cd _build/default && /Users/mac1000/.opam/5.0.0/bin/ocamlopt.opt -w -40 -g -o eio/tests/fuzz.exe /Users/mac1000/.opam/5.0.0/lib/cmdliner/cmdliner.cmxa /Users/mac1000/.opam/5.0.0/lib/ocplib-endian/ocplib_endian.cmxa /Users/mac1000/.opam/5.0.0/lib/ocaml/unix/unix.cmxa /Users/mac1000/.opam/5.0.0/lib/afl-persistent/afl-persistent.cmxa /Users/mac1000/.opam/5.0.0/lib/ocaml/str/str.cmxa /Users/mac1000/.opam/5.0.0/lib/crowbar/crowbar.cmxa /Users/mac1000/.opam/5.0.0/lib/cstruct/cstruct.cmxa -I /Users/mac1000/.opam/5.0.0/lib/cstruct /Users/mac1000/.opam/5.0.0/lib/sexplib0/sexplib0.cmxa /Users/mac1000/.opam/5.0.0/lib/base/caml/caml.cmxa /Users/mac1000/.opam/5.0.0/lib/parsexp/parsexp.cmxa /Users/mac1000/.opam/5.0.0/lib/sexplib/sexplib.cmxa /Users/mac1000/.opam/5.0.0/lib/cstruct-sexp/cstruct_sexp.cmxa /Users/mac1000/.opam/5.0.0/lib/logs/logs.cmxa /Users/mac1000/.opam/5.0.0/lib/eqaf/eqaf.cmxa /Users/mac1000/.opam/5.0.0/lib/eqaf/bigstring/eqaf_bigstring.cmxa /Users/mac1000/.opam/5.0.0/lib/eqaf/cstruct/eqaf_cstruct.cmxa /Users/mac1000/.opam/5.0.0/lib/mirage-crypto/mirage_crypto.cmxa -I /Users/mac1000/.opam/5.0.0/lib/mirage-crypto /Users/mac1000/.opam/5.0.0/lib/hkdf/hkdf.cmxa /Users/mac1000/.opam/5.0.0/lib/mirage-crypto-rng/mirage_crypto_rng.cmxa /Users/mac1000/.opam/5.0.0/lib/zarith/zarith.cmxa -I /Users/mac1000/.opam/5.0.0/lib/zarith /Users/mac1000/.opam/5.0.0/lib/mirage-crypto-pk/mirage_crypto_pk.cmxa /Users/mac1000/.opam/5.0.0/lib/ptime/ptime.cmxa /Users/mac1000/.opam/5.0.0/lib/asn1-combinators/asn1_combinators.cmxa /Users/mac1000/.opam/5.0.0/lib/fmt/fmt.cmxa /Users/mac1000/.opam/5.0.0/lib/gmap/gmap.cmxa /Users/mac1000/.opam/5.0.0/lib/domain-name/domain_name.cmxa /Users/mac1000/.opam/5.0.0/lib/base64/base64.cmxa /Users/mac1000/.opam/5.0.0/lib/mirage-crypto-ec/mirage_crypto_ec.cmxa -I /Users/mac1000/.opam/5.0.0/lib/mirage-crypto-ec /Users/mac1000/.opam/5.0.0/lib/pbkdf/pbkdf.cmxa /Users/mac1000/.opam/5.0.0/lib/macaddr/macaddr.cmxa /Users/mac1000/.opam/5.0.0/lib/ipaddr/ipaddr.cmxa /Users/mac1000/.opam/5.0.0/lib/x509/x509.cmxa /Users/mac1000/.opam/5.0.0/lib/ppx_sexp_conv/runtime-lib/ppx_sexp_conv_lib.cmxa /Users/mac1000/.opam/5.0.0/lib/ipaddr-sexp/ipaddr_sexp.cmxa /Users/mac1000/.opam/5.0.0/lib/tls/tls.cmxa /Users/mac1000/.opam/5.0.0/lib/hmap/hmap.cmxa /Users/mac1000/.opam/5.0.0/lib/lwt-dllist/lwt_dllist.cmxa /Users/mac1000/.opam/5.0.0/lib/optint/optint.cmxa /Users/mac1000/.opam/5.0.0/lib/eio/core/eio__core.cmxa /Users/mac1000/.opam/5.0.0/lib/bigstringaf/bigstringaf.cmxa -I /Users/mac1000/.opam/5.0.0/lib/bigstringaf /Users/mac1000/.opam/5.0.0/lib/mtime/mtime.cmxa /Users/mac1000/.opam/5.0.0/lib/eio/eio.cmxa /Users/mac1000/.opam/5.0.0/lib/ptime/clock/os/ptime_clock.cmxa -I /Users/mac1000/.opam/5.0.0/lib/ptime/clock/os eio/tls_eio.cmxa /Users/mac1000/.opam/5.0.0/lib/psq/psq.cmxa /Users/mac1000/.opam/5.0.0/lib/eio/utils/eio_utils.cmxa /Users/mac1000/.opam/5.0.0/lib/eio/mock/eio_mock.cmxa /Users/mac1000/.opam/5.0.0/lib/logs/logs_fmt.cmxa eio/tests/.fuzz.eobjs/native/dune__exe.cmx eio/tests/.fuzz.eobjs/native/dune__exe__Mock_socket.cmx eio/tests/.fuzz.eobjs/native/dune__exe__Test_rng.cmx eio/tests/.fuzz.eobjs/native/dune__exe__Fuzz.cmx)
# ld: warning: directory not found for option '-L/opt/local/lib'
# File "eio/tests/tls_eio.md", line 1, characters 0-0:
# /usr/local/bin/git --no-pager diff --no-index --color=always -u _build/default/eio/tests/tls_eio.md _build/default/eio/tests/.mdx/tls_eio.md.corrected
# diff --git a/_build/default/eio/tests/tls_eio.md b/_build/default/eio/tests/.mdx/tls_eio.md.corrected
# index dace2d7..7ecfa1e 100644
# --- a/_build/default/eio/tests/tls_eio.md
# +++ b/_build/default/eio/tests/.mdx/tls_eio.md.corrected
# @@ -106,11 +106,5 @@ let serve_ssl ~config server_s callback =
#         test_client ~net ("127.0.0.1", "4433")
#      )
#    ;;
# -+server -> start @ tcp:127.0.0.1:4433
# -+server -> connect
# -+handler accepted
# -+handler + GET / HTTP/1.1
# -+client <- GET
# -+client done.
# -- : unit = ()
# +Exception: Eio.Io Eio_luv.Luv_error(EPERM) (* operation not permitted *)
#  ```
# (cd _build/default/eio/tests && ./fuzz.exe --repeat 200)
# random ops: PASS
```

**Comment by hannesm** [comment]
Revdep faliures (fixed by applying upper bounds)
- dns https://github.com/mirage/ocaml-dns/pull/331 //cc @hannesm @reynir
- albatross //cc @hannesm @reynir 
- paf (+paf-cohttp) //cc @dinosaure https://github.com/dinosaure/paf-le-chien/pull/84
- letters https://github.com/oxidizing/letters/pull/53 @joseferben
- sihl //cc @joseferben 
- mehari-mirage //cc @ArtichOwO
- sendmail-lwt //cc @dinosaure https://github.com/mirage/colombe/pull/67
- capnp-rpc-unix //cc @talex5 (Mirage_crypto_rng_unix.initialize has a different signaure, see #23282
- ocaml-irc-client //cc @johnelse https://github.com/johnelse/ocaml-irc-client/pull/61
- tlstunnel-lwt //cc @hannesm https://github.com/hannesm/tlstunnel-lwt/pull/30

If you need help to update your project, please don't hesitate to reach out.

This PR looks fine to be merged now :)

**Comment by dinosaure** [comment]
The error on MacOS seems unrelated to this release, let's merge :)

**Pull request merged**